Accept a `child_index` lambda on `fields_for` with an argument

keegnotrub opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

While `fields_for` currently accepts a lambda as an argument for `child_index`, that lambda is called without any arguments. Being that we are enumerating the association on `fields_for` at the time the lambda is called, we have an opportunity to pass the child of the association down to the lambda.  This could simplify the case where you may want to use the `id` of an Active Record object as the index.  So with this change you could do something like this:

```html.erb
<%= author_form.fields_for :books, child_index: ->(book) { book.id } do |book_form| %>
  <%= book_form.text_field :name %>
<% end %>
```

Prior to this, you would have had to enumerate the association yourself:

```html.erb
<% author.books.each do |book| %>
  <%= author_form.fields_for :books, book, child_index: book.id do |book_form| %>
    <%= book_form.text_field :name %>
  <% end %>
<% end %>
```

### Detail

Accepts passing the association's child as an argument in a lambda as the `child_index` option in `fields_for`.

### Additional information

Used [PR 19661](https://github.com/rails/rails/pull/19661) as a guide when passing a lambda to `child_index` was originally introduced.



<!-- Provide additional information such as benchmarks, reference to other repositories or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

keegnotrub committed:
Accept a `child_index` lambda on `fields_for` with an argument

Accepts passing the association's child as an argument in a lambda as
the `child_index` option in `fields_for`.

Co-authored-by: Stefanni Brasil <stefannibrasil@gmail.com>
Co-authored-by: Sean Doyle <sean.p.doyle24@gmail.com>
Co-authored-by: Louis Antonopoulos <19808194+louis-antonopoulos@users.noreply.github.com>
Co-authored-by: Aji Slater <aji.slater@gmail.com>

seanpdoyle commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -2722,7 +2722,13 @@ def fields_for_with_nested_attributes(association_name, association, options, bl
             association.each do |child|
               if explicit_child_index
-                options[:child_index] = explicit_child_index.call if explicit_child_index.respond_to?(:call)
+                if explicit_child_index.respond_to?(:call)
+                  if explicit_child_index.arity.zero?
+                    options[:child_index] = explicit_child_index.call
+                  else
+                    options[:child_index] = explicit_child_index.call(child)
+                  end
+                end
```
Could this set of branches be flattened based on the arity?

```suggestion
                if explicit_child_index.respond_to?(:call)
                  arguments = [child].take(explicit_child_index.arity)
                  options[:child_index] = explicit_child_index.call(*arguments)
                end
```

seanpdoyle replied:
Or, if that's too clever:

```suggestion
                if explicit_child_index.respond_to?(:call)
                  options[:child_index] =
                    if explicit_child_index.arity.zero?
                      explicit_child_index.call
                    else
                      explicit_child_index.call(child)
                    end
                end
```

keegnotrub replied:
Your first suggestion make sense, but to be honest I did have to verify the following in `irb`

```ruby
> child_index = -> { "hi" }
> child_index.call(*[])
=> "hi"
```

keegnotrub committed:
flatten `if` branching on `child_index` lambda check

both branches set `options[:child_index]` so we just set it as the result of the conditional instead.

Co-authored-by: Sean Doyle <seanpdoyle@users.noreply.github.com>
